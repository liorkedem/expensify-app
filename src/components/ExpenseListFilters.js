import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        }
        else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-gorup">
                    <span className="input-gorup__item">
                        <input
                            type="text"
                            className="text-input"
                            value={this.props.filters.text}
                            onChange={this.onTextChange} />
                    </span>
                    <span className="input-gorup__item">
                        <select
                        className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}>
                            <option value="date">date</option>
                            <option value="amount">amount</option>
                        </select>
                    </span>
                    <span className="input-gorup__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </span>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);