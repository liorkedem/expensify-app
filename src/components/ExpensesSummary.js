import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';


export class ExpensesSummary extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        const expensesTerm = this.props.expensesCount === 1 ? 'expense' : 'expenses';

        return (
            <div>
                <p>
                    Viewing {this.props.expensesCount} {expensesTerm} totalling {numeral(this.props.expensesTotal / 100).format('$0,0.00')}
                </p>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(
        state.expenses,
        state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);