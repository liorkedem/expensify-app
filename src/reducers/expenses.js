const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(exp => exp.id !== action.expense.id);
        case 'EDIT_EXPENSE':
            return state.map((exp) => {
                if (exp.id === action.id) {
                    return {
                        ...exp,
                        ...action.expense.updates
                    }
                } else {
                    return exp;
                }
            });
        default:
            return state;
    }
};
