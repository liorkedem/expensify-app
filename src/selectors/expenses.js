import moment from 'moment';

export default (expenses, filters) => {
    return expenses.filter((exp) => {
        const createdAtMoment = moment(exp.createdAt);
        const startDateMatch = filters.startDate ? filters.startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = filters.endDate ? filters.endDate.isSameOrAfter(createdAtMoment, 'day'): true;

        const textMatch = exp.description.toLowerCase().includes(filters.text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
        if (filters.sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else      if (filters.sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

