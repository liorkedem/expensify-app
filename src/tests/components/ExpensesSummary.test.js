import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={1000} />);
    expect(wrapper).toMatchSnapshot();
});


test('should render ExpensesSummary with 3 expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={2000} />);
    expect(wrapper).toMatchSnapshot();
});