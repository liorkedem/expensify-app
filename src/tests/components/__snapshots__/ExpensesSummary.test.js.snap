// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $10.00
  </p>
</div>
`;

exports[`should render ExpensesSummary with 3 expense 1`] = `
<div>
  <p>
    Viewing 
    2
     
    expenses
     totalling 
    $20.00
  </p>
</div>
`;
